### DISCUSSION TITLE
Running NPT in NequIP

### DISCUSSION THREAD

**Discussion by sinaDFT** [mir-group/nequip#309] [General]
Hello,

I have trained a NequIP model based on the NPT data at zero pressure from AIMD VASP and I have tested it for some unseen data and the acurracy is good. Now, I am wondering if I can deploy the model and run NPT in LAMMPS or ASE using NequIP? Is NPT run is possible on NequIP currently? and if the NPT run is not implemented yet, can I use the model trained on the VASP NPT data (p=0) and run NVT using NequIP?

**Comment by Linux-cpp-lisp** [comment]
Hi @sinaDFT ,

Yes, there is stress support for `pair_nequip` in beta on https://github.com/mir-group/pair_nequip/tree/stress.

You can use a model trained with stresses to run NPT, definitely.

**Reply by sinaDFT** [reply]
@Linux-cpp-lisp Currently, I have trained the nequip model based on the energy and forces from NPT vasp data. So, if I train the model on virial stresses, I should basically train the model using force and virial data, right? for the loss fucntion, I have the norms for force and stress instead of force and energy? Can you please elaborate more about this or direct me to the previous discussions that have covered this part or any example in nequip repo about this? Thank you

**Reply by Linux-cpp-lisp** [reply]
> for the loss fucntion, I have the norms for force and stress instead of force and energy?

what do you mean?

**Reply by sinaDFT** [reply]
I mean, the usual case for the loss function is the weighted sum of energy and a force loss terms. But, when I train on forces and stresses, the loss function would be the weighted sum of stress and a force loss terms (no energy loss term)?

**Reply by Linux-cpp-lisp** [reply]
Well, you certainly can train on only forces and stress, but you can also have a weighted sum of all three in the loss, which probably makes the most sense if you have data for all three.

**Reply by sinaDFT** [reply]
@Linux-cpp-lisp Thank you for your response. I actually went ahead and tried to train the model based on the weighted sum of forces, total_energy and virial (stress) and I got the following error:

`RuntimeError: Loss function include fields virial that are not predicted by the model {'pos': 1x1o, 'edge_index': None, 'node_attrs': 1x0e, 'node_features': 16x0e, 'edge_attrs': 1x0e+1x1o+1x2e+1x3o, 'edge_embedding': 8x0e, 'atomic_energy': 1x0e, 'total_energy': 1x0e, 'forces': 1x1o}`

In the `.yaml` file, for the key_mapping aand loss_coeffs, I have added the virial part as:


key_mapping:
  z: atomic_numbers                                                            
  E: total_energy                                                                  
  F: forces                                                                       
  R: pos
  pbc: pbc
  cell: cell
  virial: virial
----------------------------------------------------------------
loss_coeffs:                                                                      
  forces: 
    - 1.0
  virial:
    - 1.0                                                                        
  total_energy:                                                                    
    - 1.0
    - PerAtomMSELoss
----------------------------------------------------------------

**Reply by Linux-cpp-lisp** [reply]
Please see `minimal_stress.yaml` and see that at present you have to include `StressForceOutput` in your `model_builders`

**Reply by sinaDFT** [reply]
@Linux-cpp-lisp Thank you.

**Comment by fxcoudert** [comment]
Hi, slightly different question here. We have a model trained with the main branch of nequip, which was trained on data that did not include stress. The final model behaves correctly in NVT MD in LAMMPS, and the value of pressure is reasonable. However, simulations performed in the NPT ensemble leak to the box basically exploding. Even though the model was not trained on stress, it can be calculated by virial from the atomic forces. Is that not the case? Is `pair_nequip` master branch currently not handling NPT simulations?

We have looked at the `stress` branch, but it does a different job: it uses the predictor to calculate stress, which is not what we want in our case.

I guess another way to phrase the question is: in LAMMPS with current `pair_nequip` master branch, where does the pressure in NVT come from?

**Reply by fxcoudert** [reply]
I'm looking at that line:

```
  // TODO: Virial stuff? (If there even is a pairwise force concept here)
```

and I really think it should instead be:

```
  virial_fdotr_compute();
```

In the current state of things, the virial is never set. How would an NPT simulation work in this case? Is there another place in LAMMPS that actually sets the virial? Most other potentials (except many-body and some Intel accelerators) seem to do it there. I've looked at the content of `virial_fdotr_compute()` and it seems to do what we would expect.

**Reply by Linux-cpp-lisp** [reply]
Hi @fxcoudert --- pressures/stress/virial are completely wrong on `pair_nequip` `main` because they are not set, so the virial is just zeros;  even if you didn't train on stress you should use the `stress` branch.  Sorry, still in the middle of sorting out all the merging down here...

But yes as you say you can compute the virial without training on it, although you can't do it using fdotr. You just have to include `StressForceOutput` instead of `ForceOutput` in your `model_builders`.

**Comment by fxcoudert** [comment]
@Linux-cpp-lisp thanks for the response, there are a couple of points that are unclear to me:

> even if you didn't train on stress you should use the stress branch

What stress will be produced with the `stress` branch and a model that was trained without stresses?
I'm looking at the pair_nequip code in that branch and it pulls the stress tensor from the model 
https://github.com/mir-group/pair_nequip/blob/97f3f7b20c02e964e78365c5c0854faec5ee2480/pair_nequip.cpp#L478

But if the model was trained without stress, how does it predict it?

**Comment by youngjae142** [comment]
@fxcoudert 
Lammps stress include two terms: kinetic and virial
I think what main branch calculate is kinetic term which doesnt need interaction info (I guess).

**Comment by Linux-cpp-lisp** [comment]
Yes @ssrokyz is correct (thanks!);  this is why the results on the `main` branch are nonsense for your calculation.

> But if the model was trained without stress, how does it predict it?

The stress can be predicted through autodifferentiation of the energy regardless of whether stress labels were used in training. Usually they are more accurate if stress labels are included, and for some applications it has been demonstrated that including the labels is critical, but you can still compute decent stress labels from a decent model trained only on energies and forces.  (Intuitively, note that forces give extremely similar gradient information about the potential energy surface to stresses/virials.)

**Comment by sinaDFT** [comment]
@Linux-cpp-lisp So, I am wondering if there is some difference between training the nequip with NPT AIMD data and NVT AIMD data? The thing is when I am training the network with forces and energies from NPT AIMD, the validation error and test error for both energy and forces are very small and nequip has really good predictions. However, when I train the network with NVT AIMD data the force prediction is perfect but energy predictions are so bad. I played around with number of data for train, weights of the force and energy terms in loss functions and some other hyperparameter tuning and none of them works. Important question is that why with traning on NVT AIMD data, I cannot have good prediction like NPT AIMD data?

**Reply by Linux-cpp-lisp** [reply]
Is your test set sampled from NPT in both cases?

**Reply by sinaDFT** [reply]
Actually, this is the process:
Training on NPT data and testing on NPT data
Training on NVT data and testing on NVT data

**Reply by sinaDFT** [reply]
@Linux-cpp-lisp Any idea why in the case of NVT, the model cannot have good predictions for energy? Thanks.

**Reply by youngjae142** [reply]
Is it possible that stress training helps energy training?
It can be the case, considering that stress is derivative of energy (they are dependent on each other).
I mean, NPT data provide better stress data, and so on.

**Reply by sinaDFT** [reply]
@ssrokyz I train the Nequip on NPT energies and forces without considering the stress/virial and I get good results for energy and force predictions. However, if I train the Nequip on NVT energy and forces, I get good predictions for force and very bad predictions for energies. So, I guess it is independent from stress/virial.

**Reply by simonbatzner** [reply]
Hi @sinaDFT, can you quantify "very bad"? This seems odd, NVT/NPT should have little to no effect on how hard energies are to learn if well equilibrated. Most of our tests have been on data sampled from NVT and consistently show <1meV/atom energy MAEs.
